File: app/src/main/java/com/afollestad/photoaffix/dialogs/ImageSizingDialog.java
package com.afollestad.photoaffix.dialogs;

import android.app.Activity;
import android.app.Dialog;
import android.app.DialogFragment;
import android.graphics.Bitmap;
import android.os.Bundle;
import android.support.v7.app.AppCompatActivity;
import android.view.View;
import android.widget.AdapterView;
import android.widget.ArrayAdapter;
import android.widget.SeekBar;
import android.widget.Spinner;
import android.widget.TextView;

import com.afollestad.materialdialogs.MaterialDialog;
import com.afollestad.photoaffix.R;
import com.afollestad.photoaffix.ui.MainActivity;

import butterknife.BindView;
import butterknife.ButterKnife;
import butterknife.Unbinder;

/** @author Aidan Follestad (afollestad) */
public class ImageSizingDialog extends DialogFragment {

  @BindView(R.id.inputWidth)
  TextView inputWidth;
  @BindView(R.id.inputHeight)
  TextView inputHeight;
  @BindView(R.id.inputScaleSeek)
  SeekBar inputScaleSeek;
  @BindView(R.id.inputScaleLabel)
  TextView inputScaleLabel;
  @BindView(R.id.formatSpinner)
  Spinner formatSpinner;
  @BindView(R.id.qualityTitle)
  TextView qualityTitle;
  @BindView(R.id.qualitySeeker)
  SeekBar qualitySeeker;
  @BindView(R.id.qualityLabel)
  TextView qualityLabel;
  private SizingCallback callback;
  private int minimum;
  private Unbinder unbinder;

  public ImageSizingDialog() {}

  public static void show(AppCompatActivity context, int width, int height) {
    ImageSizingDialog dialog = new ImageSizingDialog();
    Bundle args = new Bundle();
    args.putInt("width", width);
    args.putInt("height", height);
    dialog.setArguments(args);
    dialog.show(context.getFragmentManager(), "[IMAGE_SIZING_DIALOG]");
  }

  private double round(double x) {
    return Math.round(x * 100.0) / 100.0;
  }

  @Override
  public Dialog onCreateDialog(Bundle savedInstanceState) {
    MaterialDialog dialog =
        new MaterialDialog.Builder(getActivity())
            .title(R.string.output_settings)
            .customView(R.layout.dialog_imagesizing, false)
            .positiveText(R.string.continueStr)
            .negativeText(android.R.string.cancel)
            .neutralText(R.string.defaultStr)
            .autoDismiss(false)
            .onPositive(
                (materialDialog, dialogAction) -> {
                  if (callback != null) {
                    final int progress = inputScaleSeek.getProgress() + minimum;
                    final double scale = round((double) progress / 1000d);
                    callback.onSizingResult(
                        scale,
                        Integer.parseInt(inputWidth.getText().toString().trim()),
                        Integer.parseInt(inputHeight.getText().toString().trim()),
                        formatSpinner.getSelectedItemPosition() == 0
                            ? Bitmap.CompressFormat.PNG
                            : Bitmap.CompressFormat.JPEG,
                        qualitySeeker.getProgress() + 1,
                        false);
                  }
                  MainActivity.dismissDialog(materialDialog);
                })
            .onNegative(
                (materialDialog, dialogAction) -> {
                  callback.onSizingResult(0.0f, -1, -1, Bitmap.CompressFormat.PNG, 100, true);
                  MainActivity.dismissDialog(materialDialog);
                })
            .onNeutral(
                (materialDialog, dialogAction) ->
                    inputScaleSeek.setProgress(inputScaleSeek.getMax()))
            .build();

    final View v = dialog.getCustomView();
    assert v != null;
    unbinder = ButterKnife.bind(this, v);

    minimum = (int) (inputScaleSeek.getMax() * 0.1d);
    inputScaleSeek.setMax((int) (inputScaleSeek.getMax() * 0.9d));
    inputScaleSeek.setOnSeekBarChangeListener(
        new SeekBar.OnSeekBarChangeListener() {
          @Override
          public void onProgressChanged(SeekBar seekBar, int progress, boolean fromUser) {
            progress += minimum;
            final double scale = round((double) progress / 1000d);
            String scaleStr;
            if (scale == 0) scaleStr = "0.00";
            else if (scale == 1) scaleStr = "1.00";
            else {
              scaleStr = String.format("%s", scale);
              if (scaleStr.length() == 3) scaleStr += "0";
            }
            inputScaleLabel.setText(scaleStr);
            final int originalWidth = getArguments().getInt("width");
            final int originalHeight = getArguments().getInt("height");
            inputWidth.setText(String.format("%d", (int) (originalWidth * scale)));
            inputHeight.setText(String.format("%d", (int) (originalHeight * scale)));
          }

          @Override
          public void onStartTrackingTouch(SeekBar seekBar) {}

          @Override
          public void onStopTrackingTouch(SeekBar seekBar) {}
        });
    inputScaleSeek.setProgress(inputScaleSeek.getMax());

    ArrayAdapter<String> adapter =
        new ArrayAdapter<>(getActivity(), R.layout.spinner_item, new String[] {"PNG", "JPEG"});
    adapter.setDropDownViewResource(R.layout.spinner_item_dropdown);
    formatSpinner.setOnItemSelectedListener(
        new AdapterView.OnItemSelectedListener() {
          @Override
          public void onItemSelected(AdapterView<?> parent, View view, int position, long id) {
            final int visibility = position == 1 ? View.VISIBLE : View.GONE;
            qualityTitle.setVisibility(visibility);
            qualitySeeker.setVisibility(visibility);
            qualityLabel.setVisibility(visibility);
            qualitySeeker.setProgress(99);
          }

          @Override
          public void onNothingSelected(AdapterView<?> parent) {}
        });
    formatSpinner.setAdapter(adapter);

    qualitySeeker.setOnSeekBarChangeListener(
        new SeekBar.OnSeekBarChangeListener() {
          @Override
          public void onProgressChanged(SeekBar seekBar, int progress, boolean fromUser) {
            qualityLabel.setText(String.format("%d", progress + 1));
          }

          @Override
          public void onStartTrackingTouch(SeekBar seekBar) {}

          @Override
          public void onStopTrackingTouch(SeekBar seekBar) {}
        });

    qualitySeeker.setMax(99);
    qualitySeeker.setProgress(99);

    return dialog;
  }

  @Override
  public void onAttach(Activity activity) {
    super.onAttach(activity);
    if (activity instanceof SizingCallback) {
      callback = (SizingCallback) activity;
    }
  }

  @Override
  public void onDestroyView() {
    super.onDestroyView();
    unbinder.unbind();
    unbinder = null;
  }

  public interface SizingCallback {
    void onSizingResult(
        double scale,
        int resultWidth,
        int resultHeight,
        Bitmap.CompressFormat format,
        int quality,
        boolean cancelled);
  }
}

File: dialogs/src/main/java/com/afollestad/photoaffix/dialogs/ImageSizingDialog.kt
/*
 * Licensed under Apache-2.0
 *
 * Designed and developed by Aidan Follestad (@afollestad)
 */
package com.afollestad.photoaffix.dialogs

import android.app.Dialog
import android.content.Context
import android.content.DialogInterface
import android.graphics.Bitmap.CompressFormat
import android.graphics.Bitmap.CompressFormat.JPEG
import android.graphics.Bitmap.CompressFormat.PNG
import android.os.Bundle
import android.view.View.GONE
import android.view.View.VISIBLE
import android.widget.ArrayAdapter
import androidx.appcompat.app.AppCompatActivity
import androidx.fragment.app.DialogFragment
import com.afollestad.materialdialogs.MaterialDialog
import com.afollestad.materialdialogs.customview.customView
import com.afollestad.materialdialogs.customview.getCustomView
import com.afollestad.photoaffix.utilities.ext.onItemSelected
import com.afollestad.photoaffix.utilities.ext.onProgressChanged
import com.afollestad.photoaffix.utilities.ext.round
import kotlinx.android.synthetic.main.dialog_imagesizing.view.formatSpinner
import kotlinx.android.synthetic.main.dialog_imagesizing.view.inputHeight
import kotlinx.android.synthetic.main.dialog_imagesizing.view.inputScaleLabel
import kotlinx.android.synthetic.main.dialog_imagesizing.view.inputScaleSeek
import kotlinx.android.synthetic.main.dialog_imagesizing.view.inputWidth
import kotlinx.android.synthetic.main.dialog_imagesizing.view.qualityLabel
import kotlinx.android.synthetic.main.dialog_imagesizing.view.qualitySeeker
import kotlinx.android.synthetic.main.dialog_imagesizing.view.qualityTitle

interface SizingCallback {
  fun onSizeChanged(
    scale: Double,
    resultWidth: Int,
    resultHeight: Int,
    format: CompressFormat,
    quality: Int,
    cancelled: Boolean
  )
}

/** @author Aidan Follestad (afollestad) */
class ImageSizingDialog : DialogFragment() {

  companion object {
    private const val TAG = "[IMAGE_SIZING_DIALOG]"

    fun show(
      context: AppCompatActivity,
      width: Int,
      height: Int
    ) {
      val dialog = ImageSizingDialog()
      val args = Bundle().apply {
        putInt("width", width)
        putInt("height", height)
      }
      dialog.arguments = args
      dialog.show(context.supportFragmentManager, TAG)
    }
  }

  private var callback: SizingCallback? = null
  private var minimum: Int = 0

  override fun onCreateDialog(savedInstanceState: Bundle?): Dialog {
    val context = activity
    requireNotNull(context)

    val dialog = MaterialDialog(context)
        .title(R.string.output_settings)
        .customView(R.layout.dialog_imagesizing)
        .positiveButton(R.string.continueStr) { invokeCallback() }
        .negativeButton(android.R.string.cancel) {
          callback?.onSizeChanged(
              scale = 0.0,
              resultWidth = -1,
              resultHeight = -1,
              format = PNG,
              quality = 100,
              cancelled = true
          )
          dismiss()
        }
        .noAutoDismiss()

    val customView = dialog.getCustomView() ?: return dialog
    minimum = (customView.inputScaleSeek.max * 0.1).toInt()

    customView.inputScaleSeek.max = (customView.inputScaleSeek.max * 0.9).toInt()
    customView.inputScaleSeek.onProgressChanged { oldProgress ->
      var newProgress = oldProgress
      newProgress += minimum

      val scale = (newProgress.toDouble() / 1000.0).round()
      val scaleStr = when (scale) {
        0.0 -> "0.00"
        1.0 -> "1.00"
        else -> scale.toString().apply {
          if (this.length == 3) this.plus("0")
        }
      }
      customView.inputScaleLabel.text = scaleStr

      val originalWidth = arguments?.getInt("width") ?: -1
      val originalHeight = arguments?.getInt("height") ?: -1
      customView.inputWidth.text = (originalWidth * scale).toInt()
          .toString()
      customView.inputHeight.text = (originalHeight * scale).toInt()
          .toString()
    }
    customView.inputScaleSeek.progress = customView.inputScaleSeek.max

    val adapter = ArrayAdapter(context, R.layout.spinner_item, arrayOf("PNG", "JPEG"))
    adapter.setDropDownViewResource(R.layout.spinner_item_dropdown)
    customView.formatSpinner.onItemSelected {
      val visibility = if (it == 1) VISIBLE else GONE
      customView.qualityTitle.visibility = visibility
      customView.qualitySeeker.visibility = visibility
      customView.qualityLabel.visibility = visibility
      customView.qualitySeeker.progress = 99
    }
    customView.formatSpinner.adapter = adapter

    customView.qualitySeeker.max = 99
    customView.qualitySeeker.onProgressChanged {
      customView.qualityLabel.text = (it + 1).toString()
    }
    customView.qualitySeeker.progress = 99

    return dialog
  }

  private fun invokeCallback() {
    val materialDialog = dialog as? MaterialDialog ?: return
    val customView = materialDialog.getCustomView() ?: return

    val progress = customView.inputScaleSeek.progress + minimum
    val scale = (progress.toDouble() / 1000.0).round()
    val width = customView.inputWidth.text.toString()
        .trim()
        .toInt()
    val height = customView.inputHeight.text.toString()
        .trim()
        .toInt()
    val quality = customView.qualitySeeker.progress + 1

    callback?.onSizeChanged(
        scale = scale,
        resultWidth = width,
        resultHeight = height,
        format = selectedFormat(),
        quality = quality,
        cancelled = false
    )
    dismiss()
  }

  override fun onAttach(context: Context?) {
    super.onAttach(context)
    callback = context as? SizingCallback
  }

  override fun onCancel(dialog: DialogInterface?) {
    super.onCancel(dialog)
    callback?.onSizeChanged(
        scale = 0.0,
        resultWidth = -1,
        resultHeight = -1,
        format = PNG,
        quality = 100,
        cancelled = true
    )
  }

  private fun selectedFormat(): CompressFormat = with(dialog as MaterialDialog) {
    val customView = getCustomView() ?: return PNG
    return if (customView.formatSpinner.selectedItemPosition == 0) {
      PNG
    } else {
      JPEG
    }
  }
}
